app/assets/javascripts/discourse/controllers/insert-hyperlink.js.es6

```diff
@@ -1,15 +1,98 @@
 import ModalFunctionality from "discourse/mixins/modal-functionality";
+import { searchForTerm } from "discourse/lib/search";
+import { observes } from "ember-addons/ember-computed-decorators";
 
 export default Ember.Controller.extend(ModalFunctionality, {
-  linkUrl: "",
-  linkText: "",
-
   onShow() {
-    Ember.run.next(() =>
-      $(this)
-        .find("input.link-url")
-        .focus()
-    );
+    this.set("linkUrl", "");
+    this.set("linkText", "");
+    this.set("searchResults", []);
+    this.set("selectedRow", -1);
+
+    Ember.run.next(() => {
+      const element = document.querySelector(".insert-link");
+
+      element.addEventListener("keydown", e => {
+        switch (e.which) {
+          case 40:
+            this.selectRow(e, "down");
+            break;
+          case 38:
+            this.selectRow(e, "up");
+            break;
+          case 13:
+            // override Enter behaviour when a row is selected
+            if (this.selectedRow > -1) {
+              this.setUpdated(this.selectedRow);
+              element.querySelector("input.link-text").focus();
+              e.preventDefault();
+              e.stopPropagation();
+            }
+            break;
+        }
+      });
+
+      element
+        .closest(".modal-inner-container")
+        .addEventListener("mousedown", e => {
+          if (!e.target.closest(".inputs")) {
+            this.set("searchResults", []);
+          }
+        });
+
+      document.querySelector("input.link-url").focus();
+    });
+  },
+
+  selectRow(e, direction) {
+    const index =
+      direction === "down" ? this.selectedRow + 1 : this.selectedRow - 1;
+
+    if (index > -1 && index < this.searchResults.length) {
+      document
+        .querySelectorAll(".internal-link-results .search-link")
+        [index].focus();
+      this.set("selectedRow", index);
+    } else {
+      this.set("selectedRow", -1);
+      document.querySelector("input.link-text").focus();
+    }
+
+    e.preventDefault();
+  },
+
+  setUpdated(index) {
+    const topic = this.searchResults[index];
+    if (topic) {
+      this.set("linkUrl", Discourse.BaseUrl + topic.url);
+      if (!this.linkText) {
+        this.set("linkText", topic.title);
+      }
+    }
+    this.set("selectedRow", -1);
+  },
+
+  @observes("linkUrl")
+  triggerSearch() {
+    if (
+      this.linkUrl &&
+      this.linkUrl.length > 3 &&
+      !this.linkUrl.startsWith("http")
```

Comment:
startsWith is not available under IE11 (https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/String/startsWith) and we decided to not polyfill it because you can do it very simply with `indexOf`.